src/bin/path_filestat.rs

```diff
@@ -0,0 +1,135 @@
+use libc;
+use misc_tests::open_scratch_directory;
+use misc_tests::utils::{cleanup_file, close_fd};
+use misc_tests::wasi::{
+    wasi_fd_fdstat_get, wasi_path_filestat_get, wasi_path_filestat_set_times, wasi_path_open,
+};
+use std::{env, process};
+
+fn test_path_filestat(dir_fd: libc::__wasi_fd_t) {
+    let mut fdstat: libc::__wasi_fdstat_t = unsafe { std::mem::zeroed() };
+    let status = wasi_fd_fdstat_get(dir_fd, &mut fdstat);
+    assert_eq!(status, libc::__WASI_ESUCCESS, "fd_fdstat_get");
+
+    assert!(
+        (fdstat.fs_rights_base & libc::__WASI_RIGHT_PATH_FILESTAT_GET) != 0,
+        "the scratch directory should have __WASI_RIGHT_PATH_FILESTAT_GET as base right",
+    );
+    assert!(
+        (fdstat.fs_rights_inheriting & libc::__WASI_RIGHT_PATH_FILESTAT_GET) != 0,
+        "the scratch directory should have __WASI_RIGHT_PATH_FILESTAT_GET as base right",
+    );
+
+    // Create a file in the scratch directory.
+    let mut file_fd = libc::__wasi_fd_t::max_value() - 1;
+    let filename = "file";
+    let status = wasi_path_open(
+        dir_fd,
+        0,
+        filename,
+        libc::__WASI_O_CREAT,
+        libc::__WASI_RIGHT_FD_READ
+            | libc::__WASI_RIGHT_FD_WRITE
+            | libc::__WASI_RIGHT_PATH_FILESTAT_GET,
+        0,
+        0,
+        &mut file_fd,
+    );
+    assert_eq!(status, libc::__WASI_ESUCCESS, "opening a file");
+    assert!(
+        file_fd > libc::STDERR_FILENO as libc::__wasi_fd_t,
+        "file descriptor range check",
+    );
+
+    let status = wasi_fd_fdstat_get(file_fd, &mut fdstat);
+    assert_eq!(status, libc::__WASI_ESUCCESS, "fd_fdstat_get");
+
+    assert_eq!(
+        fdstat.fs_rights_base & libc::__WASI_RIGHT_PATH_FILESTAT_GET,
+        0,
+        "files shouldn't have rights for path_* syscalls even if manually given",
+    );
+    assert_eq!(
+        fdstat.fs_rights_inheriting & libc::__WASI_RIGHT_PATH_FILESTAT_GET,
+        0,
+        "files shouldn't have rights for path_* syscalls even if manually given",
+    );
+
+    // Check file size
+    let mut stat = libc::__wasi_filestat_t {
+        st_dev: 0,
+        st_ino: 0,
+        st_filetype: 0,
+        st_nlink: 0,
+        st_size: 0,
+        st_atim: 0,
+        st_mtim: 0,
+        st_ctim: 0,
+    };
+    let status = wasi_path_filestat_get(dir_fd, 0, filename, filename.len(), &mut stat);
+    assert_eq!(status, libc::__WASI_ESUCCESS, "reading file stats");
+    assert_eq!(stat.st_size, 0, "file size should be 0");
+
+    // Check path_filestat_set_times
+    let old_atim = stat.st_atim;
+    let new_mtim = stat.st_mtim - 100;
+    let status = wasi_path_filestat_set_times(
+        dir_fd,
+        0,
+        filename,
+        filename.len(),
+        new_mtim,
```

Comment:
Should this not be `old_atim`?

Reply:
This is on purpose. See line 94: `atim` should not changed even I pass `new_mtim`, because the flag is not set.

Reply:
Since we don't pass `__WASI_FILESTAT_SET_ATIM`, I guess in this particular case it doesn't matter. Nonetheless, it's quite confusing when you read the code :-)

Reply:
That's fair enough. Could you add a comment to that effect then please?

Reply:
Done